Support readonly for all fields

gmazzamuto opened the pull request:
Support `:readonly` option for all fields.

The `:readonly` field option, being shared among all fields, has been moved to the `@config_schema` of `Backpex.Field`. This should not be a breaking change.

Example of readonly fields:

<img width="1216" height="954" alt="Screenshot 2025-11-25 at 01-24-43 Edit User · Backpex" src="https://github.com/user-attachments/assets/233dc76d-145d-47bb-aedc-11532935d1f9" />

gmazzamuto committed:
Support readonly for all fields

The `:readonly` field option, being shared among all fields, has been
moved to the `@config_schema` of `Backpex.Field`.

Flo0807 commented on lib/backpex/html/form.ex:
```diff
@@ -205,6 +212,13 @@ defmodule Backpex.HTML.Form do
+        assigns
+      else
+        put_in(assigns, [:rest, :disabled], Map.get(assigns.rest, :readonly) || false)
+      end
```
See https://github.com/naymspace/backpex/pull/1677/files#r2591617008

Flo0807 commented on lib/backpex/html/core_components.ex:
```diff
@@ -64,21 +65,38 @@ defmodule Backpex.HTML.CoreComponents do
+           |> List.delete("input"))
+      else
+        trigger_class
+      end
```
This looks complicated. Can't we just add additional classes in the markup in case the dropdown is readonly?

gmazzamuto replied:
Here is why. This is how the dropdown is rendered with the code above:

<img width="554" height="66" alt="Screenshot_20251223_173629" src="https://github.com/user-attachments/assets/39ecfdf8-277c-414c-a3b8-8e6e5ebc1315" />

---

Whereas this code:
```elixir
    trigger_class =
      if assigns.readonly do
        ["cursor-not-allowed bg-base-200"] ++ trigger_class
      else
        trigger_class
      end
```

results in this, and the field can be even highlighted on mouse click (although the dropdown does not open):

<img width="552" height="65" alt="Screenshot_20251223_173556" src="https://github.com/user-attachments/assets/cd77986e-5b59-4237-8bb8-0c3e8847cfe4" />

Flo0807 commented on lib/backpex/html/form.ex:
```diff
@@ -47,6 +47,13 @@ defmodule Backpex.HTML.Form do
+        assigns
+      else
+        put_in(assigns, [:rest, :disabled], Map.get(assigns.rest, :readonly) || false)
+      end
```
See https://github.com/naymspace/backpex/blob/0.16.3/lib/backpex/fields/text.ex#L58

I noticed that we do not set `dsiabled` in every field (e.g. `Backpex.Fields.Select`)

Flo0807 commented on lib/backpex/fields/multi_select.ex:
```diff
@@ -161,6 +161,7 @@ defmodule Backpex.Fields.MultiSelect do
         </:label>
         <Form.multi_select
           field={@form[@name]}
+          readonly={@readonly}
```
> undefined attribute "readonly" for component Backpex.HTML.Form.multi_select/1

Flo0807 commented on lib/backpex/html/form.ex:
```diff
@@ -286,6 +300,7 @@ defmodule Backpex.HTML.Form do
   @doc type: :component
 
   attr :prompt, :string, required: true, doc: "string that will be shown when no option is selected"
+  attr :readonly, :boolean, default: false
```
```suggestion
  attr :readonly, :boolean, default: false, doc: "whether the dropdown is readonly"
```

Flo0807 commented on lib/backpex/html/resource.ex:
```diff
@@ -175,7 +175,13 @@ defmodule Backpex.HTML.Resource do
+        assigns
+      else
+        assign(assigns, :readonly, Backpex.Field.readonly?(field_options, assigns))
+      end
```
For which cases do we need this?

And can we simplify the if-condition?


```suggestion
    assigns =
      if assigns[:readonly] do
        assigns
      else
        assign(assigns, :readonly, Backpex.Field.readonly?(field_options, assigns))
      end
```

gmazzamuto committed:
Support readonly for all fields: changes for PR review

Flo0807 committed:
Merge remote-tracking branch 'origin/develop' into feature/readonly-fields

Flo0807 committed:
Fix readonly guide link label in field.ex docstring

Flo0807 committed:
Render dropdown readonly as inert div instead of fake button

Previously the readonly dropdown kept role="button", tabindex=0, and
aria-haspopup="true" while stripping the menu, producing a focusable
element that announces a popup that isn't there. Also removed the
fragile class-token-surgery that reached into caller class lists to
delete "input" and "bg-transparent" strings; callers now build their
own readonly-aware class lists using Phoenix's list syntax.

Flo0807 committed:
Gate Upload field readonly for drop target, cancel buttons, and link

The drop target's phx-drop-target attribute was not gated on readonly,
so drag-and-drop still started uploads. Cancel buttons for in-progress
and existing entries remained clickable, letting users mutate file
state. The "Upload a file" <a> rendered as a dead link in readonly
mode — now a plain <span> with the interactive classes removed.

Flo0807 committed:
Improve readonly prompt contrast to meet WCAG AA

text-base-content/40 on bg-base-200 falls below 4.5:1 on stock
daisyUI themes. Bumped to /60 to match the codebase's standard
muted-text token (used in help_text / labels).

Flo0807 committed:
Drop redundant disabled={@readonly} from native-readonly inputs

HTML's readonly and disabled have different a11y semantics: disabled
removes the element from the tab order and is frequently skipped by
screen readers in form-review mode, making the readonly value
invisible. For inputs that natively support readonly (text, number,
date, date_time, time, textarea, email, url, currency), keeping only
readonly={@readonly} is correct. Controls without native readonly
(select, multi_select, boolean, relations, upload) continue to use
disabled as before.

Flo0807 committed:
Hide InlineCRUD delete and add controls entirely when readonly

Previously the readonly branch removed the delete checkbox but left
an orphaned <label for> pointing at nothing, wrapping a <div
class="btn btn-disabled"> with sr-only "Delete" text. Screen readers
announced a non-interactive "Delete" with a dangling label
association. Same problem for "Add entry" — rendered as a disabled
checkbox-masquerading-as-button. In readonly mode a user has no use
for either control, so they are now omitted entirely.

Flo0807 committed:
Thread readonly through has_many_through relational sub-components

select_relational_field/1 and pivot_field/1 previously did not
receive the parent field's readonly flag. The main UI already hid
the triggers that invoke them, so no live regression — but any
future path that opens the relation modal (e.g. programmatic
push_event, pre-populated newest_relational) would render an
editable select on a readonly field. Now both components declare
readonly as an explicit attr and forward it to their inputs.

Flo0807 committed:
Update readonly guide for all fields supported after the PR

The guide listed only Date, DateTime, Number, Text, Textarea as
readonly-capable. With this PR, readonly is a global Backpex.Field
option and most field types support it. Documented the three
rendering strategies (native-readonly, disabled, custom) and the
per-field quirks for Upload, InlineCRUD, HasManyThrough, Boolean.

Flo0807 committed:
Document readonly behavior in Upload, InlineCRUD, HasManyThrough moduledocs

These three fields render readonly with custom UI changes that hide
or disable elements beyond the standard readonly/disabled attribute.
Added short Readonly sections to the moduledocs so users discover
this behavior without reading the guide.

Flo0807 committed:
Add component tests for dropdown/1 and multi_select/1 readonly

Covers the readonly-branch HTML surface of both components:
dropdown has no role/tabindex/aria-haspopup/menu when readonly;
multi_select prompt uses /60 contrast and badges render without
primary color or remove controls. Sets up test/html/ as the home
for future HTML component tests (the repo previously had no
component-level coverage).

gmazzamuto committed:
Fix readonly InlineCRUD

This fixes a bug caused by Ecto's `:sort_param` and `:drop_param` when
the field is readonly. The hidden fields need to be marked as disabled
to prevent being sent when the InlineCRUD field is readonly. Otherwise
the following bug happens. Steps to reproduce:

* in UserLive, set the social_links field as readonly
* in the edit view, make any change to the form
* the social_links will disappear because Ecto's `cast_embed` doesn't
  work properly

gmazzamuto committed:
Merge branch 'develop' into feature/readonly-fields
